### PR TITLE
fix(slonik): adjust types according to slonik unit tests and documentation

### PR DESCRIPTION
**Pull request by n1ru4l**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/gajus/slonik#slonik-query-methods
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/slonik/index.d.ts [hunk 1]
```diff
@@ -221,15 +221,15 @@ export type QueryResultRowType<ColumnName extends string = string> = {
...
-export type QueryFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<QueryResultType<T>>;
-export type QueryManyFirstFunctionType = QueryAnyFirstFunctionType;
+export type QueryAnyFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<ReadonlyArray<T>>;
+export type QueryAnyFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<ReadonlyArray<T>>;
 export type QueryManyFunctionType = QueryAnyFunctionType;
-export type QueryMaybeOneFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<T[keyof T] | null>;
+export type QueryManyFirstFunctionType = QueryAnyFirstFunctionType;
 export type QueryMaybeOneFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<T | null>;
-export type QueryOneFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<T[keyof T]>;
+export type QueryMaybeOneFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<T | null>;
```
Comment (mmkal) [inline types/slonik/index.d.ts]: This isn't right. `T[keyof T]` was correct. The `...First` functions return the first column value of the row objects. Which means if `T = {foo: string}`, `T[keyof T] = string`.

File: types/slonik/index.d.ts [hunk 2]
```diff
@@ -221,15 +221,15 @@ export type QueryResultRowType<ColumnName extends string = string> = {
     [name in ColumnName]: QueryResultRowColumnType;
 };
 
-export type QueryAnyFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<Array<T[keyof T]>>;
-export type QueryAnyFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<T[]>;
-export type QueryFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<QueryResultType<T>>;
-export type QueryManyFirstFunctionType = QueryAnyFirstFunctionType;
+export type QueryAnyFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<ReadonlyArray<T>>;
+export type QueryAnyFirstFunctionType = <T>(...args: QueryMethodParams<T>) => Promise<ReadonlyArray<T>>;
```
Comment (mmkal) [inline types/slonik/index.d.ts]: These aren't readonly arrays, and there's no reason to type them that way. You lose the ability to use utility functions for regular arrays by doing this.

File: types/slonik/slonik-tests.ts [hunk 3]
```diff
@@ -110,18 +110,18 @@ const typedQuery = async () => {
   await pool.query(getFooBarQuery(10));
 
   // $ExpectType string
-  await pool.oneFirst(getFooQuery(10));
+  await pool.oneFirst<string>(getFooQuery(10));
```
Comment (mmkal) [inline types/slonik/slonik-tests.ts]: The generic in all these query methods is supposed to represent the row type, which could never be string.

**Review by mmkal — CHANGES_REQUESTED** [review]
These changes don't reflect how the library works - if there's something you're finding is going wrong for you, let me know and I can help with adding a test case, or some documentation to the types.

**Comment by n1ru4l** [comment]
@mmkal

How would you type this correctly?

```ts
const amountOfRecords: number = ((await this.slonik.oneFirst(sql`
  SELECT COUNT(*) FROM "records"
  WHERE "ownerId" = 1;
`)) as unknown) as number;
```

**Comment by mmkal** [comment]
@n1ru4l 
```typescript
const amountOfRecords = await slonik.oneFirst<{records: number}>(sql`
  SELECT COUNT(*) FROM "records"
  WHERE "ownerId" = 1;
`)
```

I created a package under the `@slonik` org which will do that automatically for you. It's very new, but feel free to try it out and see if it works for you: https://npmjs.com/package/@slonik/typegen

**Pull request closed without merging**